Border Controls: Update control components to allow 40px height

aaronrobertshaw opened the pull request:
Related:

- https://github.com/WordPress/gutenberg/issues/40893#issuecomment-1160061349

## What?

Adds the ability for the `BorderControl`, `BorderBoxControl`, and `BorderRadiusControl` components to be displayed with a `40px` height.

At this stage, this PR does not toggle the display of border controls to `40px` by default. If we wish to make these controls that height straight away we can close this PR in favor of more straightforward tweaks.

## Why?

The current design direction for new tools / controls is to have a `40px` default height.

<img src="https://user-images.githubusercontent.com/548849/174545138-bacefd5a-3095-4165-93ac-a92cb445ff71.png" width="400"/>

## How?

- Replaces the `__next36pxDefaultSize` prop for the `BorderControl` and `BorderBoxControl` components with `__next40pxDefaultSize`
- Updates the `BorderRadiusControl` to accept a new `__next40pxDefaultSize` prop
- Uses the `__next40pxDefaultSize` props to make each component display with a `40px` height
- The `BorderRadiusControl` leverages the `size` prop of its inner `UnitControl`/`InputControl` to make its input `40px`. 

## Testing Instructions

1. After checking out this PR, edit a post, add a Group block, and select it
2. Within the inspector controls sidebar the border controls should still be displayed within a 30px height
3. Switch to the site editor and within the Global Styles panel navigate to Blocks > Group > Layout
4. Confirm the border controls are still 30px
5. Apply the patch below to toggle on the 40px height for border controls
6. Ensure the border controls in both editors are now 40px high and all inner components are positioned correctly

<details>
<summary><strong>Diff to toggle on 40px height</strong></summary>

```diff
diff --git a/packages/block-editor/src/hooks/border-radius.js b/packages/block-editor/src/hooks/border-radius.js
index fa61bafa92..f062373a19 100644
--- a/packages/block-editor/src/hooks/border-radius.js
+++ b/packages/block-editor/src/hooks/border-radius.js
@@ -38,6 +38,7 @@ export function BorderRadiusEdit( props ) {
 		<BorderRadiusControl
 			values={ style?.border?.radius }
 			onChange={ onChange }
+			__next40pxDefaultSize={ true }
 		/>
 	);
 }
diff --git a/packages/block-editor/src/hooks/border.js b/packages/block-editor/src/hooks/border.js
index 81cc023bab..3766b56912 100644
--- a/packages/block-editor/src/hooks/border.js
+++ b/packages/block-editor/src/hooks/border.js
@@ -275,6 +275,7 @@ export function BorderPanel( props ) {
 						value={ hydratedBorder }
 						__experimentalHasMultipleOrigins={ true }
 						__experimentalIsRenderedInSidebar={ true }
+						__next40pxDefaultSize={ true }
 					/>
 				</ToolsPanelItem>
 			) }
diff --git a/packages/edit-site/src/components/global-styles/border-panel.js b/packages/edit-site/src/components/global-styles/border-panel.js
index 71fe697697..f9fcf1951f 100644
--- a/packages/edit-site/src/components/global-styles/border-panel.js
+++ b/packages/edit-site/src/components/global-styles/border-panel.js
@@ -186,6 +186,7 @@ export default function BorderPanel( { name } ) {
 						popoverOffset={ 40 }
 						__experimentalHasMultipleOrigins={ true }
 						__experimentalIsRenderedInSidebar={ true }
+						__next40pxDefaultSize={ true }
 					/>
 				</ToolsPanelItem>
 			) }
@@ -201,6 +202,7 @@ export default function BorderPanel( { name } ) {
 						onChange={ ( value ) => {
 							setBorderRadius( value || undefined );
 						} }
+						__next40pxDefaultSize={ true }
 					/>
 				</ToolsPanelItem>
 			) }

```
</details>

## Screenshots or screencast <!-- if applicable -->

<details>
<summary><strong>Global Styles</strong></summary>

| Before | After |
|---|---|
| <img width="318" alt="Screen Shot 2022-06-22 at 5 34 22 pm" src="https://user-images.githubusercontent.com/60436221/174979605-38edff93-3439-4b2d-853a-049ba84f2d0a.png"> | <img width="302" alt="Screen Shot 2022-06-22 at 5 30 27 pm" src="https://user-images.githubusercontent.com/60436221/174979676-94d352d9-4a4d-4e9e-875a-c136a4aeda0c.png"> |
</details>


<details>
<summary><strong>Block Editor</strong></summary>

| Before | After |
|---|---|
| <img width="297" alt="Screen Shot 2022-06-22 at 6 18 19 pm" src="https://user-images.githubusercontent.com/60436221/174979907-d00e5b03-0c3f-4866-8e0a-9fd27bbb312d.png"> |  <img width="305" alt="Screen Shot 2022-06-22 at 4 52 15 pm" src="https://user-images.githubusercontent.com/60436221/174979997-e426469d-b073-472a-86c2-4c96bdc351a7.png"> |
</details>

aaronrobertshaw committed:
Update border control components to allow 40px height

aaronrobertshaw committed:
Make 40px height work again after rebase

aaronrobertshaw committed:
Turn on 40px tall border controls in editors

aaronrobertshaw committed:
Update radius control to handle extra padding

These tweaks update the BorderRadiusControl such that the entered values are actually visible despite the increased internal padding applied to the UnitControl when the `__unstable-large` variant is used.

aaronrobertshaw committed:
First pass at improving border controls when at 40px

Initial tweaks to improve the UX for the border controls. Aims to address the increased padding for large 40px high UnitControl variant that cuts of entered values from display.

aaronrobertshaw commented:
This PR has been updated to also include turning on the 40px high input in our editors as per this [request](https://github.com/WordPress/gutenberg/issues/40893#issuecomment-1242751383).

We've recently updated the styling for large control variants within the editors to increase internal padding. For example, the `UnitControl` using the `__unstable-large` size gets increased padding on its input as well as prefix and suffix. 

This is a bit of a problem currently for the border controls as they use both a prefix and a suffix. When also constrained by the width of our sidebar there isn't much space to actually display entered values. This is accentuated further in the split view, configuring individual sides.

<details>

<summary><strong>Click for demo of above issue before tweaks</strong></summary>

https://user-images.githubusercontent.com/60436221/192202635-aee40c18-03a2-4df1-b88e-53e69c778e3d.mp4



</details>


I've taken an initial pass at trying to tweak the control styles and input widths to maximise the space available without overriding the UnitControl's padding so as to keep it in line with other controls. It's better but there's still a lot more that needs improving.

@jasmussen or @jameskoster, if you have time to cast an eye over this and provide your thoughts and suggestions, it would be greatly appreciated 🙇 

<details>

<summary><strong>Click for demo of current state in this PR</strong></summary>

https://user-images.githubusercontent.com/60436221/192202717-89465723-b03d-409a-bb41-9a2a186230d8.mp4

</details>

jameskoster commented:
Thanks for the ping @aaronrobertshaw.

I just mocked this up in Figma (aligned to the grid we've been working towards) and everything seems to fit okay:

<img width="454" alt="Screenshot 2022-09-26 at 11 17 58" src="https://user-images.githubusercontent.com/846565/192252659-9f5579cd-d3ef-4195-ac98-61d26fb0b634.png">

At a quick glance, I think we can increase the width of the inputs in the unlinked view, but ideally we should aim to fit the grid.

In case it's helpful, here's the [Figma file](https://www.figma.com/file/9wDJslLfdeDHxC1ula4lgG/Spacing?node-id=0%3A1) to inspect which should include all relevant values.

jasmussen commented:
Thank you for working on this, this is excellent attention to detail 🙏 

Jay summarizes well. I'll only add — and this may not be needed for this PR — that it'd be nice of the "Swatch box" could be square. This one:

<img width="96" alt="Screenshot 2022-09-26 at 13 27 03" src="https://user-images.githubusercontent.com/1204802/192265228-da02c469-e54c-4857-8cad-1c164f9b035d.png">

jameskoster commented:
> that it'd be nice of the "Swatch box" could be square

I thought it was, good spot 😅

aaronrobertshaw commented:
> I just mocked this up in Figma (aligned to the grid we've been working towards) and everything seems to fit okay:

Thanks for the updated mockup and figma file @jameskoster 👍 

I'll massage the component into shape today.

> that it'd be nice of the "Swatch box" could be square

This was something I'd tweaked locally but exacerbated the issue of overly constraining the display of the input value so I left it out for this first pass. It's an easy change if we are able to increase the overall input width.

aaronrobertshaw committed:
Make compact inputs 118px wide

aaronrobertshaw committed:
Increase padding to 16px between input and slider

aaronrobertshaw committed:
Make split controls gap a uniform 16px

aaronrobertshaw committed:
Update split layout for floating linked button

aaronrobertshaw committed:
Make color dropdown 40px square with 16px indicator

aaronrobertshaw committed:
Tweak input widths again to fit editor sidebar

aaronrobertshaw committed:
Make radius control match border and grid

aaronrobertshaw commented:
I've given the border control styles another pass. It's getting closer, but there are a couple of minor issues still for it to be pixel perfect.

The figma file suggests the panel is `284px` wide, whereas the sidebar in our editors is `280px`. That means using the `118px` width for inputs as per the mockup causes things to overflow by a few pixels. I reduced the input width to `116px` so they fit neatly.

Regardless of whether the inputs were `116px` or `118px`, numeric values with 3 digits or decimals get ever so slightly truncated. 

| 3 Digits | Decimal |
|---|---|
| <img width="162" alt="Screen Shot 2022-09-27 at 4 21 50 pm" src="https://user-images.githubusercontent.com/60436221/192448489-f7aa85a1-bf0e-49aa-8a29-d9d5c152b8f7.png"> | <img width="165" alt="Screen Shot 2022-09-27 at 4 22 11 pm" src="https://user-images.githubusercontent.com/60436221/192448486-601e720f-1a5f-47a6-8929-3b5362a335c2.png"> | 

The `Mixed` placeholder text gets more truncated after these latest changes so you now only see `Mix`. 

<img width="161" alt="Screen Shot 2022-09-27 at 4 22 39 pm" src="https://user-images.githubusercontent.com/60436221/192448476-5fab950e-b837-4a78-aab4-25c1fade0990.png">

I looked at removing the unit select when the border control has a mixed set of borders however this felt rather odd when it was just a color difference causing the mixed state. The control was designed so even if the colors were different you could still adjust the width value and unit as desired. 

The border radius control has also been updated so it appears more in line with the main border control.

Here's a quick video of the current state of these controls.


https://user-images.githubusercontent.com/60436221/192449806-324a125d-3669-4559-b72a-3efd09fa5e2c.mp4



Any thoughts on ways to improve the controls from here?

jameskoster commented:
For the truncation issue, I think the best solution would be to remove the padding on the unit button. Doing so ensures 24x24px dimensions which are acceptable, and creates just enough space for `rem`, the longest unit:

<img width="282" alt="Screenshot 2022-09-27 at 10 35 45" src="https://user-images.githubusercontent.com/846565/192491638-b6fb535b-11f6-4be5-99a8-9ff0d0a29751.png">

Ideally that wouldn't be an override for just this control. Do we need a new button variant? It doesn't look like we have a simple no-padding, no-border option 🤔 

This doesn't quite create enough room for the 'Mixed' value though. Must it say 'Mix**ed**', or is 'Mix' adequate? That fits:

<img width="280" alt="Screenshot 2022-09-27 at 10 42 44" src="https://user-images.githubusercontent.com/846565/192492860-af93be18-dab7-4596-b137-3a81a3a35347.png">

The only other option I can think of would be to hide the unit input when the values are mixed... it doesn't seem to work properly at the moment anyway 🙈

jameskoster commented on packages/block-editor/src/components/border-radius-control/style.scss:
```diff
@@ -11,9 +11,9 @@
 
 		> .components-unit-control-wrapper {
-			width: 110px;
+			width: 116px;
```
I don't know if it's possible but ideally we shouldn't use absolute values like this. The controls should instead fill the width of the container, and the space between be determined via `gap`. This means that if we ever increase/decrease the width of the sidebar then everything will update gracefully.

aaronrobertshaw replied:
Thanks for flagging this 👍

We can avoid the hardcoded value for the `BorderRadiusControl` and so I've updated it in https://github.com/WordPress/gutenberg/pull/41860/commits/0f2b87c0a4af4f7cbfe67648a0eccc5567769db4. The `BorderBoxControl` is a little more complicated so I think that might be better addressed in a follow-up.

aaronrobertshaw committed:
Remove hardcoded width from radius input wrapper

aaronrobertshaw committed:
Try using Mix as placeholder for mixed values

aaronrobertshaw commented:
Appreciate the continued feedback @jameskoster 🙇 

> For the truncation issue, I think the best solution would be to remove the padding on the unit button. Doing so ensures 24x24px dimensions which are acceptable, and creates just enough space for rem, the longest unit:

I haven't pushed the removal of padding from the unit select button yet as I got caught up exploring the potential removal of the unit select button when we have mixed values. That said, it does provide just enough breathing room for those longer values. The plan is to also strip the unit select's padding from the radius control as well so at least the two controls that are next to each other in that panel match up.

> This doesn't quite create enough room for the 'Mixed' value though. Must it say 'Mixed', or is 'Mix' adequate?

I'm not sure if 'Mix' is adequate, it seems like a poorer option to me than 'Mixed', however, I think we are short of alternatives at this point. At least in terms of placeholders. On top of that, languages other than English will likely struggle to find something meaningful that is 3 characters or less.

> when the values are mixed... it doesn't seem to work properly at the moment anyway 🙈

The `BorderBoxControl` and `BorderRadiusControl` do behave slightly differently when mixed values are present. I have something in the works here that should bring them closer regarding functionality while still allowing some of the other border properties to be updated as originally intended.

That would be better as a separate PR focused on those changes independent of the visual tweaks here. I'll publish a PR for it tomorrow.

jameskoster commented:
> I think we are short of alternatives at this point

Could another option be a wildcard label, IE "*"? I'm not sure how intuitive that would be though. Perhaps with a tooltip it's be ok?

<img width="200" alt="Screenshot 2022-09-28 at 10 59 40" src="https://user-images.githubusercontent.com/846565/192750622-5193fdec-9d09-46b0-9c21-fcd319bdbfe6.png">

Another thought... does the linked control really need to support mixed values? What if toggling away from unlinked borders re-linked them automatically?

aaronrobertshaw commented:
> Could another option be a wildcard label, IE "*"? I'm not sure how intuitive that would be though. Perhaps with a tooltip it's be ok?

That's an interesting idea. It might not be 100% intuitive but I don't know that it is any worse than "Mix" or whatever non-English placeholders get truncated to. 

The addition of the tooltip might be a little bit trickier. I think we'd need to be able to apply the tooltip to the UnitControl's inner input as we already have a tooltip on the color/style dropdown. Without being able to target the inner input we'd end up with multiple tooltips when the color and style dropdown was hovered over.

There's a draft PR up that explores adding a tooltip for an `InputControl`'s inner input. That was put on hold until we could sort out a broader approach via https://github.com/WordPress/gutenberg/issues/42630.


> Another thought... does the linked control really need to support mixed values? What if toggling away from unlinked borders re-linked them automatically?

I think it might cause some friction if this control behaves differently than those around it. The spacing controls for padding, margin etc, box controls, and the border radius control, all allow that mixed state without "re-linking" the values. 

I'm also aware of users relying on that link button to simply collapse the control in the sidebar so it doesn't take up so much space. Now the inspector controls sidebar primarily uses the `ToolsPanel`; there isn't yet a means of collapsing the panel or removing a control from display without resetting it. 

There's been discussion around adding a way to collapse ToolsPanels but they tend to fizzle out due to crowding the panel with buttons or obscure behaviour.

---

Instead of a tooltip, could we leverage some help text? 

This might provide a way to;

- communicate the mixed state,
- be similar to the 'Mixed' text added to spacing control labels,
- provide extra space for other languages,
- avoid issues with multiple tooltips

From an a11y point of view, might it also provide an improvement over the simple placeholder?

| <img width="277" alt="Screen Shot 2022-09-29 at 6 18 43 pm" src="https://user-images.githubusercontent.com/60436221/192982013-58045f59-3854-44ef-bcee-5324b29b8f6e.png"> | <img width="277" alt="Screen Shot 2022-09-29 at 6 20 56 pm" src="https://user-images.githubusercontent.com/60436221/192982007-d144cae4-4a29-4a34-b5d6-d6a5db565e3a.png"> |
|---|---|

jasmussen commented:
Thanks for all the work here, this is a strong improvement!

About the "Mixed" issue, I would avoid showing values in helper text. The asterisk could work. But I would think the strongest solution is to hide the "px" unit and show the regular "Mixed" text in the input (the units can be mixed too, afterall). Even if we have to elide the text to "Mix.", or simply allow it to be cropped by the input field, that still feels the most clear and simple. 

That's not to say we can't come back to this with stronger solutions, at a later point. It may also be possible to earn some space back by exploring a design that omits the separator between border indicator and value:

<img width="623" alt="Screenshot 2022-09-29 at 13 49 39" src="https://user-images.githubusercontent.com/1204802/193024151-edb37194-1746-451a-9f9e-dc75eb2756dd.png">

But that's for a different time!

ciampo commented:
This is a good one for @mirka to check once she's back from AFK

aaronrobertshaw committed:
Revert placeholder back to 'Mixed'

aaronrobertshaw commented:
Thank you to everyone for all the feedback, thoughts, and ideas 🙇 

I've switched the border controls back to using the 'Mixed' placeholder (https://github.com/WordPress/gutenberg/pull/41860/commits/962a853f8b627ef750e950f6e0575ef65f14ff9e). I'll create a separate PR removing the unit select when there are mixed values.

jasmussen reviewed (approved):
Tall and graceful. This is a substantial improvement, thank you:

![40px](https://user-images.githubusercontent.com/1204802/193214674-1a352e1e-8814-4620-a6fe-9351f1155fc0.gif)

We can probably massage some paddings inside the unit button and we can even explore smaller paddings when just for the border control if need be. But this works very well for the majority of cases, rem-to-px comma conversaion notwithstanding 😅

aaronrobertshaw commented:
Thanks for taking the time to review this again @jasmussen 🙇 

> Tall and graceful. This is a substantial improvement, thank you:

I'm glad it's a step in the right direction!

> This is a good one for @mirka to check once she's back from AFK

I'd still like to give @mirka the chance to be able to weigh in on everything this PR entails if we can. Including best approaches to cleanly handle internal padding within large controls.

Given that, I'll hold off merging for the moment.

>  But this works very well for the majority of cases, rem-to-px comma conversaion notwithstanding

It appears that the comma in question is the decimal comma for your locale, correct?

There has been no change in how units get switched for border widths here. By that, I mean when the unit changes, it is just appended to the current width's numeric value. This is the current behaviour on trunk and is derived from the `UnitControl` so it is the same across this component as well as `BoxControl`, `BorderRadiusControl` etc.

If this is an issue, it might be something we'll need to address closer to the source in the `UnitControl` component.

jasmussen commented:
> It appears that the comma in question is the decimal comma for your locale, correct?

Oh, right. And yes I guess you can add decimals. But I think the meta point to make was that you can type in any crazy big number, and it will fill more than the input. That's fine, and how inputs have always worked, so long as it works well for most cases.

ciampo commented on packages/block-editor/src/components/border-radius-control/style.scss:
```diff
@@ -27,31 +27,39 @@
-			margin-right: $grid-unit-10;
+			.component-border-radius-control__linked-button.has-icon {
+				margin-top: 8px;
+			}
```
Just a reminder that we should not use any internal classnames from the components package (i.e. `.components-*`, `.has-icon`...) — it would be great if we could work on a (low priority) follow-up to refactor these styles.

aaronrobertshaw replied:
Thanks, I was just reusing the selectors that were already floating around this component. I don't think it is too hard to refactor away from them if we can allow a custom classname on the UnitControl wrapper to avoid potentially brittle CSS. 

I've done this in https://github.com/WordPress/gutenberg/pull/41860/commits/a97e5624ea677d8d39d668e48a9cb73a1bd26c71 if you're interested or have other ideas.

ciampo replied:
Nice!

It also poses the interesting question on what is the assumption for the `className` prop: should it be the classname of the wrapper component (in `UnitControl`, the outer wrapper), or should it be the classname for the "main" component (in `UnitControl`, the inner input wrapper)  ?

Should this reasoning apply to all components in `@wordpress/components` ? (cc @mirka)

mirka replied:
I usually see the `className` apply to the outer wrapper. I think that convention makes sense, since `className` is mostly for (at least somewhat hacky) internal style overrides, and scoping it to the outermost wrapper allows for the most flexibility.

ciampo replied:
Agreed with Lena, and I'd be up for applying that reasoning to `UnitControl` too

But unfortunately `UnitControl` already applies `className` to the inner input wrapper — changing the prop to apply to the outer wrapper may cause some breaking changes unfortunately

ciampo replied:
Let's rename `wrapperClassName` to `__unstableWrapperClassName`, so that this PR can move forward. We will take a look at potentially ironing out this inconsistency in `UnitControl` separately

ciampo replied:
Thank you for applying those changes!

Opened #45094 to track this follow-up task

ciampo commented on packages/components/src/border-box-control/stories/index.js:
```diff
@@ -84,12 +84,12 @@ Default.args = {
 
 const WrapperView = styled.div`
-	max-width: 280px;
+	max-width: 248px;
```
How is this value computed / derived? We should probably leave a comment about it.

But in case it's a number derived from the sidebar's width, we should try to remove this "soft" dependency from this component (since components in `@wordpress/components` shouldn't know what the sidebar is), and likely let the parent component dictate the width (see, for example, https://github.com/WordPress/gutenberg/pull/44559)

aaronrobertshaw replied:
From memory, this was to help better demonstrate the space available when used within the sidebar. The component can adjust to whatever width, though, without this tweak, you had to load the editor to see the component in the wild and constrained by our sidebar's width and panel padding.

I can happily remove this change. Personally, I'm confused as to why one arbitrary width for demonstration purposes is worse than another. At the least, this helped save time while iterating on these components so it has served its purpose 🙂

ciampo replied:
>  Personally, I'm confused as to why one arbitrary width for demonstration purposes is worse than another.

My point here is that we should not have a `max-width` at all — that makes easier for someone interacting with the component on Storybook to understand its responsive behavior.

> At the least, this helped save time while iterating on these components so it has served its purpose 🙂

Storybook should offer a few tools to make it easier to resize the canvas and test the component's responsiveness

aaronrobertshaw replied:
> and likely let the parent component dictate the width (see, for example, https://github.com/WordPress/gutenberg/pull/44559)

FWIW, I think that example is a little different in that the max-width was within the component's styles. The `BorderBoxControl` does let the parent component dictate the width, which is what the `WrapperView` in the story was doing so we can see the impacts of the large input with limited width like in the sidebar.

> Storybook should offer a few tools to make it easier to resize the canvas and test the component's responsiveness

This sounds like a good follow-up, adding some tools to resize the canvas and remove the max-widths across multiple components in the storybook. I don't think it should block this PR though.

ciampo replied:
Storybook already offers some tools to resize the canvas:

https://user-images.githubusercontent.com/1083581/196409167-be7b9a68-fd56-4ce3-b645-23f9e3a66e53.mp4


Although I agree that this is not blocking.

aaronrobertshaw replied:
Thanks. I should have been clearer. I meant tools to resize to some meaningful and useful preset sizes, e.g. our sidebar widths. The device sizes don't really help, and manually drag resizing the canvas is inaccurate, especially when trying to resolve 1 or 2 pixel spacing issues.

Personally, I find it cumbersome to have to be resizing the canvas constantly and a poorer reflection on the components when they are stretched out of realistic proportions. I understand the idea that removing the restriction illustrates how the components adapt. For me, though, I think we lose as much as we gain. I'm not trying to change our direction here just highlight some of the friction experienced when using the storybook for development, not documentation.

ciampo replied:
Absolutely! This kind of feedback is incredibly valuable to us, so keep it coming!

As you mentioned, it's hard to strike the right balance between documenting and developing — I hope that setting the correct expectations with our users and that adopting a consistent approach throughout all examples (e.g. all components have unconstrained width?) can be a good direction to take.

(cc @mirka)

mirka replied:
My first choice would've been to add a "WP sidebar" preset to the [Viewport tool](https://storybook.js.org/docs/react/essentials/viewport), but that's not what we want here since changing the viewport itself would affect the font size in text fields.

So instead we'll need to add another toolbar add-on to wrap the stories in a div with some preset max-widths. This is super easy to build, so let me know if that would work for you.

aaronrobertshaw replied:
> So instead we'll need to add another toolbar add-on to wrap the stories in a div with some preset max-widths. This is super easy to build, so let me know if that would work for you.

That sounds like a good compromise. 

I took the liberty of having a go at creating the toolbar add-on ( https://github.com/WordPress/gutenberg/pull/45134 ). Let me know if I missed anything. If that PR is suitable, I'll go through our components and remove those ad hoc wrappers in a separate follow-up.

ciampo replied:
With https://github.com/WordPress/gutenberg/pull/45134 merged, I guess this wrapper view is not needed anymore?

aaronrobertshaw replied:
I intended to remove all these wrappers across multiple components in a follow-up PR. Happy to quickly strip it from here now to get the ball rolling, though 🙂

aaronrobertshaw replied:
✅ Done: https://github.com/WordPress/gutenberg/pull/41860/commits/072f6afb4905784b572ab4c3e4a576bee067bb93

ciampo replied:
Sorry, I didn't understand that was the plan! Thank you anyway!

aaronrobertshaw replied:
All good. I should have re-iterated my plans after the conversation evolved. 

I've put up a quick PR removing wrappers on non-border controls: https://github.com/WordPress/gutenberg/pull/45305.

ciampo commented on packages/components/src/border-box-control/styles.ts:
```diff
@@ -16,16 +15,23 @@ export const BorderBoxControl = css``;
 
 export const LinkedBorderControl = css`
 	flex: 1;
+	${ rtl( { marginRight: '24px' }, { marginLeft: '24px' } )() }
```
The `BorderBoxControl` component doesn't look right for RTL languages:

<img width="301" alt="Screenshot 2022-10-12 at 10 52 01" src="https://user-images.githubusercontent.com/1083581/195297168-7245c254-4faf-4d13-9c17-146697e84b6f.png">

This is due to a few reasons:

- the `LinkedBorderControl` styles are currently a constant value, meaning that even when the component re-renders, those styles never get recomputed (and therefore the `rtl` function never gets called). To fix this, we should make `LinkedBorderControl` a function that returns those styles
- the `BorderBoxControlLinkedButton` doesn't use the `rtl` function at all, and therefore the button is always positioned absolutely to the right
- minor nit: the `rtl` function can auto-magically "translate" `right` to `left` CSS props (and viceversa) without setting it explicitly
- minor nit: we should name style functions/objects with an initial lowercase letter, so that we can differentiate them at a glance from actual Styled Components (e.g. `styled.div` ...)
- as a consequence of these changes, I noticed that using `rtl.watch()` is not necessary anymore!

<details>

<summary>I prepped a diff that applies the suggestions above:</summary>

```
diff --git a/packages/components/src/border-box-control/border-box-control-linked-button/hook.ts b/packages/components/src/border-box-control/border-box-control-linked-button/hook.ts
index 61ae0ddf85..248c15b28f 100644
--- a/packages/components/src/border-box-control/border-box-control-linked-button/hook.ts
+++ b/packages/components/src/border-box-control/border-box-control-linked-button/hook.ts
@@ -25,7 +25,7 @@ export function useBorderBoxControlLinkedButton(
 	const cx = useCx();
 	const classes = useMemo( () => {
 		return cx(
-			styles.BorderBoxControlLinkedButton( __next40pxDefaultSize ),
+			styles.borderBoxControlLinkedButton( __next40pxDefaultSize ),
 			className
 		);
 	}, [ className, cx, __next40pxDefaultSize ] );
diff --git a/packages/components/src/border-box-control/border-box-control-split-controls/hook.ts b/packages/components/src/border-box-control/border-box-control-split-controls/hook.ts
index 39ba7db82a..1a9d3186b4 100644
--- a/packages/components/src/border-box-control/border-box-control-split-controls/hook.ts
+++ b/packages/components/src/border-box-control/border-box-control-split-controls/hook.ts
@@ -8,7 +8,7 @@ import { useMemo } from '@wordpress/element';
  */
 import * as styles from '../styles';
 import { useContextSystem, WordPressComponentProps } from '../../ui/context';
-import { useCx, rtl } from '../../utils/';
+import { useCx } from '../../utils/';
 
 import type { SplitControlsProps } from '../types';
 
@@ -20,25 +20,20 @@ export function useBorderBoxControlSplitControls(
 
 	// Generate class names.
 	const cx = useCx();
-	const rtlWatchResult = rtl.watch();
 	const classes = useMemo( () => {
 		return cx(
 			styles.borderBoxControlSplitControls( __next40pxDefaultSize ),
 			className
 		);
-		// rtlWatchResult is needed to refresh styles when the writing direction changes
-		// eslint-disable-next-line react-hooks/exhaustive-deps
-	}, [ cx, className, rtlWatchResult, __next40pxDefaultSize ] );
+	}, [ cx, className, __next40pxDefaultSize ] );
 
 	const centeredClassName = useMemo( () => {
-		return cx( styles.CenteredBorderControl, className );
+		return cx( styles.centeredBorderControl, className );
 	}, [ cx, className ] );
 
 	const rightAlignedClassName = useMemo( () => {
 		return cx( styles.rightBorderControl(), className );
-		// rtlWatchResult is needed to refresh styles when the writing direction changes
-		// eslint-disable-next-line react-hooks/exhaustive-deps
-	}, [ cx, className, rtlWatchResult ] );
+	}, [ cx, className ] );
 
 	return {
 		...otherProps,
diff --git a/packages/components/src/border-box-control/border-box-control-visualizer/hook.ts b/packages/components/src/border-box-control/border-box-control-visualizer/hook.ts
index cab7937f3f..cc5b47e1d4 100644
--- a/packages/components/src/border-box-control/border-box-control-visualizer/hook.ts
+++ b/packages/components/src/border-box-control/border-box-control-visualizer/hook.ts
@@ -8,7 +8,7 @@ import { useMemo } from '@wordpress/element';
  */
 import * as styles from '../styles';
 import { useContextSystem, WordPressComponentProps } from '../../ui/context';
-import { useCx, rtl } from '../../utils';
+import { useCx } from '../../utils';
 
 import type { VisualizerProps } from '../types';
 
@@ -24,15 +24,12 @@ export function useBorderBoxControlVisualizer(
 
 	// Generate class names.
 	const cx = useCx();
-	const rtlWatchResult = rtl.watch();
 	const classes = useMemo( () => {
 		return cx(
 			styles.borderBoxControlVisualizer( value, __next40pxDefaultSize ),
 			className
 		);
-		// rtlWatchResult is needed to refresh styles when the writing direction changes
-		// eslint-disable-next-line react-hooks/exhaustive-deps
-	}, [ cx, className, value, __next40pxDefaultSize, rtlWatchResult ] );
+	}, [ cx, className, value, __next40pxDefaultSize ] );
 
 	return { ...otherProps, className: classes, value };
 }
diff --git a/packages/components/src/border-box-control/border-box-control/hook.ts b/packages/components/src/border-box-control/border-box-control/hook.ts
index ce7a6f1a8a..d8db33cb29 100644
--- a/packages/components/src/border-box-control/border-box-control/hook.ts
+++ b/packages/components/src/border-box-control/border-box-control/hook.ts
@@ -97,15 +97,15 @@ export function useBorderBoxControl(
 
 	const cx = useCx();
 	const classes = useMemo( () => {
-		return cx( styles.BorderBoxControl, className );
+		return cx( styles.borderBoxControl, className );
 	}, [ cx, className ] );
 
 	const linkedControlClassName = useMemo( () => {
-		return cx( styles.LinkedBorderControl );
+		return cx( styles.linkedBorderControl() );
 	}, [ cx ] );
 
 	const wrapperClassName = useMemo( () => {
-		return cx( styles.Wrapper );
+		return cx( styles.wrapper );
 	}, [ cx ] );
 
 	return {
diff --git a/packages/components/src/border-box-control/styles.ts b/packages/components/src/border-box-control/styles.ts
index 357e14cc28..90f9b309aa 100644
--- a/packages/components/src/border-box-control/styles.ts
+++ b/packages/components/src/border-box-control/styles.ts
@@ -11,31 +11,31 @@ import { COLORS, CONFIG, rtl } from '../utils';
 import type { Border } from '../border-control/types';
 import type { Borders } from './types';
 
-export const BorderBoxControl = css``;
+export const borderBoxControl = css``;
 
-export const LinkedBorderControl = css`
+export const linkedBorderControl = () => css`
 	flex: 1;
-	${ rtl( { marginRight: '24px' }, { marginLeft: '24px' } )() }
+	${ rtl( { marginRight: '24px' } )() }
 `;
 
-export const Wrapper = css`
+export const wrapper = css`
 	position: relative;
 `;
 
-export const BorderBoxControlLinkedButton = (
+export const borderBoxControlLinkedButton = (
 	__next40pxDefaultSize?: boolean
 ) => {
 	return css`
 		position: absolute;
 		top: ${ __next40pxDefaultSize ? '8px' : '3px' };
-		right: 0;
+		${ rtl( { right: 0 } )() }
 		flex: 0;
 		flex-basis: 24px;
 		line-height: 0;
 	`;
 };
 
-const BorderBoxStyleWithFallback = ( border?: Border ) => {
+const borderBoxStyleWithFallback = ( border?: Border ) => {
 	const {
 		color = COLORS.gray[ 200 ],
 		style = 'solid',
@@ -60,13 +60,13 @@ export const borderBoxControlVisualizer = (
 		right: ${ __next40pxDefaultSize ? '39px' : '30px' };
 		bottom: ${ __next40pxDefaultSize ? '20px' : '15px' };
 		left: ${ __next40pxDefaultSize ? '39px' : '30px' };
-		border-top: ${ BorderBoxStyleWithFallback( borders?.top ) };
-		border-bottom: ${ BorderBoxStyleWithFallback( borders?.bottom ) };
+		border-top: ${ borderBoxStyleWithFallback( borders?.top ) };
+		border-bottom: ${ borderBoxStyleWithFallback( borders?.bottom ) };
 		${ rtl( {
-			borderLeft: BorderBoxStyleWithFallback( borders?.left ),
+			borderLeft: borderBoxStyleWithFallback( borders?.left ),
 		} )() }
 		${ rtl( {
-			borderRight: BorderBoxStyleWithFallback( borders?.right ),
+			borderRight: borderBoxStyleWithFallback( borders?.right ),
 		} )() }
 	`;
 };
@@ -79,11 +79,11 @@ export const borderBoxControlSplitControls = (
 	width: ${ __next40pxDefaultSize ? undefined : '80%' };
 `;
 
-export const CenteredBorderControl = css`
+export const centeredBorderControl = css`
 	grid-column: span 2;
 	margin: 0 auto;
 `;
 
 export const rightBorderControl = () => css`
-	${ rtl( { marginLeft: 'auto' }, { marginRight: 'auto' } )() }
+	${ rtl( { marginLeft: 'auto' } )() }
 `;
```

</details>

ciampo replied:
One more thing that I forgot to check during my review: we should also make sure that the popover that opens to select the border color behaves correctly / as expected for RTL languages too

aaronrobertshaw replied:
Thank you for the diff. It certainly makes life easier. Applied those changes via https://github.com/WordPress/gutenberg/pull/41860/commits/034293361110f8275e14e6a1653c478677084412

ciampo commented on packages/components/src/border-box-control/styles.ts:
```diff
@@ -16,16 +15,23 @@ export const BorderBoxControl = css``;
+	__next40pxDefaultSize?: boolean
 ) => {
 	return css`
+		position: absolute;
+		top: ${ __next40pxDefaultSize ? '8px' : '3px' };
+		right: 0;
```
Out of curiosity, any particular reason why we're using `position: absolute` and hardcoded values for vertical alignment, instead of leveraging flexbox?

Also, I think that absolute positioned elements are ignored by flexbox parents — in that case, we should be able to remove the `flex` and `flex-basis` rules below

aaronrobertshaw replied:
This used to be styled leveraging the flex box until a [revised design](https://github.com/WordPress/gutenberg/pull/41860#issuecomment-1257815773) required it to be positioned absolutely so that the split border controls for individual sides could stretch all the way to the right of the container.

I'll clean up the flex styles if they aren't required. If you have a better approach to achieving the design's layout, I'd be happy to implement it.

ciampo replied:
I see, thank you for the context — absolute positioning is definitely the way to fo when the sides are unlinked.

When the sides are linked, you could go for flexbox but I see the convenience for keeping the absolute positioning.

ciampo commented on packages/components/src/border-control/border-control/hook.ts:
```diff
@@ -118,13 +117,16 @@ export function useBorderControl(
-	const wrapperWidth = isCompact ? '90px' : width;
+	let wrapperWidth = width;
+	if ( isCompact ) {
+		wrapperWidth = __next40pxDefaultSize ? '116px' : '90px';
```
Similarly to the comment above, it would probably help if there was a comment explaining the context for these hardcoded values.

aaronrobertshaw replied:
✅ Done.

ciampo commented on packages/components/src/border-control/styles.ts:
```diff
@@ -125,13 +129,8 @@ export const colorIndicatorWrapper = (
-						width: 16px;
-				  `
-				: '' }
+			height: 16px;
```
Nit: for dimensions that match the 4px grid (including spacing and gaps), we should probably try to use the `space()` utility in order to make it clear that we're following a grid?

aaronrobertshaw replied:
I've updated these occurrences to use the `space()` util. Thanks.

ciampo commented on packages/block-editor/src/components/border-radius-control/index.js:
```diff
@@ -40,13 +45,18 @@ const MAX_BORDER_RADIUS_VALUES = {
+export default function BorderRadiusControl( {
+	onChange,
+	values,
+	__next40pxDefaultSize = false,
```
It looks like the only place left where `BorderRadiusControl` is not using `40px` inputs is in [global styles](https://github.com/WordPress/gutenberg/blob/6a2fceacb8c03a184ad22d054868bdd800b351f6/packages/edit-site/src/components/global-styles/border-panel.js)

Is the plan to migrate `BorderRadiusControl` to use `40px` height also in global styles? As soon as that is done, we should be able to remove the `__next40pxDefaultSize` prop (and undo a lot of the code changes in this PR)

aaronrobertshaw replied:
Good point. I've taken the liberty of migrating the Global Styles border-radius control to 40px height as well, then cleaned up the changes in this PR. https://github.com/WordPress/gutenberg/pull/41860/commits/cb64469a29a03bad4b4ef0113ffd0a10b9996561

ciampo replied:
Great! Much cleaner, IMO

ciampo commented on packages/components/src/border-box-control/styles.ts:
```diff
@@ -46,14 +52,14 @@ const BorderBoxStyleWithFallback = ( border?: Border ) => {
+		top: ${ __next40pxDefaultSize ? '20px' : '15px' };
+		right: ${ __next40pxDefaultSize ? '39px' : '30px' };
+		bottom: ${ __next40pxDefaultSize ? '20px' : '15px' };
+		left: ${ __next40pxDefaultSize ? '39px' : '30px' };
```
For some reason, the left border is not perfectly aligned on my screen for the old `30px` height.

Applying this change fixes it:

```suggestion
		left: ${ __next40pxDefaultSize ? '39px' : '29px' };
```

ciampo reviewed (commented):
I guess from the components' package standpoint, the main API change is the renaming of `__next36pxDefaultSize` to `__next40pxDefaultSize` for the `BorderControl` and the `BorderBoxControl` components.

From what I understand, so far we've introduced the `40px` height as a new larger size, while we have been working on setting 36px as the new default size. This PR seems to take a different approach, and instead of introducing a new "size", changes the default size from 36px to 40px. I'll let @mirka chime in as she has been working on this in the past weeks and she has by far the best understanding of the situation.

(Sidenote: In case we went ahead with this approach, I guess it's ok to remove the `__next36pxDefaultSize`, as it would at most cause the affected components to be rendered back at their `30px` height?)

mirka commented:
Tagging this PR as part of #41973

---

Sorry for holding you all up!

Short answer: API-wise, we want to go with `size = "__unstable-large"` instead of `__next*`.

Basically we got new guidance from Matias, and the `__next36pxDefaultSize` thing (#39397) is pretty much in limbo right now. It is possible that we'll scrap it entirely, although it's still unknown what the ultimate solution will be. The package-wide size consistency problem has been de-prioritized, while we instead focus on #41973.

jasmussen commented:
> From what I understand, so far we've introduced the 40px height as a new larger size, while we have been working on setting 36px as the new default size.

I know there's been some back and forth on this, sorry about the headaches. The latest is that we are settling on 40px for everything, but with a few edgecases here and there. Right now buttons in the top toolbar, for example, are a mix of 32px and 36px. One idea, **and I want to emphasize that isn't quite settled yet**, is to have 32px as the edgecase, and 40px by default:

<img width="1247" alt="strawman" src="https://user-images.githubusercontent.com/1204802/195525990-7cda4ca5-8b41-423b-92ec-82cc6dbd402b.png">

CC: @mtias @jameskoster

aaronrobertshaw committed:
Remove use of internal component classnames

aaronrobertshaw committed:
Remove unnecessary flex styles from linked button

aaronrobertshaw committed:
Use space util for color swatch dimensions

aaronrobertshaw committed:
Add explanatory comment for compact widths

aaronrobertshaw committed:
Better align the left border for 30px height variant

aaronrobertshaw committed:
Consolidate border radius controls to single 40px height

This allows us to remove most of the changes that would have been introduced in this PR as well as simplify the CSS.

aaronrobertshaw committed:
Fix RTL rendering and style naming

aaronrobertshaw committed:
Remove unneeded prop setting 40px height

aaronrobertshaw committed:
Switch __next40pxDefaultSize to size __unstable-large

aaronrobertshaw commented:
Thank you as always for the attention to detail and feedback on your reviews @ciampo and @mirka. 🙇 

I believe I've addressed most of the issues and questions. Let me know what I've missed.

> Short answer: API-wise, we want to go with size = "__unstable-large" instead of __next*.

The `__next40pxDefaultSize` has been replaced with a `size` prop in https://github.com/WordPress/gutenberg/pull/41860/commits/10d5bf90f0a1c0d4d6d82a17646b87b2b61875e5. I'll need to give it a fresh test and self-review in the morning. That said, it should be pretty close.

ciampo commented on packages/components/src/border-control/types.ts:
```diff
@@ -119,12 +119,11 @@ export type BorderControlProps = ColorProps &
+		 * Size of the control.
 		 *
-		 * @default false
+		 * @default default
```
```suggestion
		 * @default 'default'
```
I believe the README also needs to be updated

ciampo commented on packages/components/src/border-control/types.ts:
```diff
@@ -161,12 +160,11 @@ export type DropdownProps = ColorProps & {
+	 * Size of the control.
 	 *
-	 * @default false
+	 * @default default
```
```suggestion
	 * @default 'default'
```

I believe the README also needs to be updated

ciampo commented on packages/components/src/unit-control/types.ts:
```diff
@@ -100,4 +100,8 @@ export type UnitControlProps = Omit< UnitSelectControlProps, 'unit' > &
+		/**
+		 * Custom CSS class to apply to the unit control's outer wrapper.
+		 */
+		wrapperClassName?: string;
```
Same thing re. the README

ciampo commented on packages/components/src/border-box-control/types.ts:
```diff
@@ -44,12 +44,11 @@ export type BorderBoxControlProps = ColorProps &
+		 * Size of the control.
 		 *
-		 * @default false
+		 * @default default
```
Same as above for all the `'default'` values and the README updating in this file

ciampo commented on packages/components/src/border-box-control/stories/index.js:
```diff
@@ -84,12 +84,12 @@ Default.args = {
 		width: '1px',
 	},
-	__next36pxDefaultSize: false,
+	size: 'default',
```
Not for this PR, but we should refactor `BorderControl` and `BorderBoxControl` stories to TypeScript and controls, since at the moment there's no easy way to set the size to `'__unstable-large'` without typing it.

aaronrobertshaw replied:
@mirka is already one step ahead and has already converted the `BorderControl` stories to TypeScript and Controls a couple of months ago. I'll get a PR together to bring `BorderBoxControl` up-to-date as a follow-up to this.

aaronrobertshaw replied:
A quick refactor of the `BorderBoxControl` stories to TypeScript etc is up: https://github.com/WordPress/gutenberg/pull/45002.

ciampo reviewed (commented):
Thank you for addressing the feedback so far!

Let's double-check with @mirka if the new APIs are looking good from her standpoint too!

We'll need to add a couple of CHANGELOG entries for the API changes to `BorderBoxControl`, `BorderControl` and `UnitControl`.

aaronrobertshaw committed:
Fix up defaults in types and update READMEs

aaronrobertshaw committed:
Add components changelog

mirka commented:
API looks good, thanks for the meticulous work here!

aaronrobertshaw commented:
Given all feedback has been addressed or moved to follow-ups (i.e. removing max-width wrappers from components, stories to TypeScript), I think we just need a final approval before I'll merge this PR.

jameskoster reviewed (approved):
Looking good visually!

| Before | After | 
| --- | --- |
| <img width="283" alt="Screenshot 2022-10-18 at 10 42 15" src="https://user-images.githubusercontent.com/846565/196396021-9bd3fe55-d00c-45cd-b9ce-f093f3eef238.png"> | <img width="285" alt="Screenshot 2022-10-18 at 10 42 23" src="https://user-images.githubusercontent.com/846565/196396067-8c2e9189-d979-4573-9163-b070f881181b.png"> |

aaronrobertshaw committed:
Rename new prop `__unstableWrapperClassName`

ciampo reviewed (approved):
Great work here! All blocking feedback has been addressed, and the remaining follow-up tasks have been defined 🚀

aaronrobertshaw committed:
UnitControl: Normalize wrapper classnames by removing outer wrapper (#45139)

aaronrobertshaw commented:
I've merged in https://github.com/WordPress/gutenberg/pull/45139, rebased to resolve conflicts, and updated the changelog to include `UnitControl` refactor normalizing class name application.

ciampo commented on packages/block-editor/src/components/border-radius-control/style.scss:
```diff
@@ -10,24 +10,21 @@
 
-			.components-range-control__wrapper {
-				margin-right: 10px;
+			> div {
+				height: 40px;
+				display: flex;
+				align-items: center;
```
just flagging this for future work — `RangeControl` doesn't yet offer a `40px` variant (cc @mirka )

ciampo reviewed (commented):
I think we're almost ready to merge!

- is the plan for #45002 to merge it into this PR, or to merge it separately ?
- all feedback except https://github.com/WordPress/gutenberg/pull/41860#discussion_r1004129506 has been addressed.

aaronrobertshaw committed:
BorderBoxControl: Refactor stories to TypeScript and Controls (#45002)

aaronrobertshaw committed:
Remove max-width wrappers from border control stories

aaronrobertshaw commented:
> is the plan for https://github.com/WordPress/gutenberg/pull/45002 to merge it into this PR, or to merge it separately ?
> all feedback except https://github.com/WordPress/gutenberg/pull/41860#discussion_r1004129506 has been addressed.

#45002 has been merged into this PR and the max-width wrappers have been stripped from the border control stories. Other components, I'll address in a follow-up soon.

I've also given this a smoke test after the recent merges and last-minute tweaks.

✅ Stories function as expected without the max-width wrapper
✅ No typing errors
✅ Unit tests are passing
✅ Controls still work in the editor

Thanks @ciampo for the continued help and the reminder regarding #45002 👍

ciampo commented:
No worries! One last rebase should fix the failing PHP unit tests :)

ciampo committed:
Merge branch 'trunk' into update/border-controls-to-40px-height

ciampo commented:
@aaronrobertshaw happy to merge this PR today in case it gets late for you

aaronrobertshaw commented:
> happy to merge this PR today in case it gets late for you

Thanks for handling the merge. The tests were still running when I had to log off for the day.

Great to see it merged when logging in 🎉
